### ISSUE TITLE
Different truth rectangle defined and used ( dnn_mmod_ex )

### ISSUE THREAD

**Issue by AdoHaha** [davisking/dlib#481]
Hi, when playing around with dnn_mmod_ex and modyfying the detection window I found an issue where I define different detector width but the 40x40 (default?) rectangle is still there. Is there a limit on the detector window size?

The program output:

detection window width,height:      14,43
overlap NMS IOU thresh:             0.775
overlap NMS percent covered thresh: 1
step#: 3510  learning rate: 0.1   average loss: 0.333093     steps without apparent progress: 101
Encountered a truth rectangle with a width and height of 16 and 42.
The image pyramid and sliding window can't output a rectangle of this shape. 
This is because the rectangle is smaller than the detection window which has a width
and height of 40 and 40.

**Comment by davisking** [comment]
Hmm, that's not what should happen.   Please post a complete program that
reproduces the problem and I'll take a look.

**Comment by AdoHaha** [comment]
I have modified only one line in the example, regarding the mmod_options, to (I have tried many different values but the 40x40 persists) mmod_options options(face_boxes_train, 18*55);

Currently, all my rectangles have 18x55 ratio. 

The output of the program right now is:

Encountered a truth rectangle with a width and height of 22 and 51.
The image pyramid and sliding window can't output a rectangle of this shape. 
This is because the rectangle's aspect ratio is too different from the detection window,
which has a width and height of 40 and 40.

 I have pasted the whole program here: http://pastebin.com/jKqAvtVg

**Comment by davisking** [comment]
Post everything necessary to run it.  If I can't run it I can't debug it.

**Comment by davisking** [comment]
I ran it and this is the output I saw:

```
num training images: 204
num testing images:  24
detection window width,height:      18,55
overlap NMS IOU thresh:             0.148148
overlap NMS percent covered thresh: 0.276923
step#: 0     learning rate: 0.1   average loss: 0            steps without apparent progress: 0
Encountered a truth rectangle with a width and height of 63 and 91.
The image pyramid and sliding window can't output a rectangle of this shape. 
This is because the rectangle's aspect ratio is too different from the detection window,
which has a width and height of 18 and 55.
```

Which is different from what you posted.  I don't know why that is.  You must not be running the current version of dlib or you have changes to the code you didn't post.  In any case, *this* error is happening because there are training boxes that have a very different aspect ratio.  In particular, your average aspect ratio is apparently 55/18, but the message complains that there is a training box being passed to the trainer that has an aspect ratio of 91/63, which is quite different.  

You should pop the images and boxes coming out of the random cropper onto the screen and make sure they look right to you.  You should probably rotate the boxes less.  That's likely why this is happening.  


As an aside, your training data is very easy and would probably work well with a HOG detector.  HOG runs a lot faster than the DNN stuff, so you should probably try that first.

**Comment by kumarvis** [comment]
Code : http://codepad.org/NQwKRE81
Details: https://stackoverflow.com/questions/44966493/mmod-dlib-object-detector-an-impossible-set-of-object-labels-was-detected

exception thrown:

exception thrown!
An impossible set of object labels was detected. This is happening because none
of the object locations checked by the supplied image scanner is a close enough
match to one of the truth boxes. To resolve this you need to either lower the 
match_eps or adjust the settings of the image scanner so that it hits this 
truth box. Or you could adjust the offending truth rectangle so it can be 
matched by the current image scanner. Also, if you are using the 
scan_image_pyramid object then you could try using a finer image pyramid or 
adding more detection templates. E.g. if one of your existing detection 
templates has a matching width/height ratio and smaller area than the offending
rectangle then a finer image pyramid would probably help. 

image index              0
match_eps:               0.5
best possible match:     0.188406
truth rect:              [(600, 448) (617, 465)]
truth rect width/height: 1
truth rect area:         324
nearest detection template rect:              [(609, 443) (626, 460)]
nearest detection template rect width/height: 1
nearest detection template rect area:         324

Some insight about the exception what it is trying to say will be great looking at the different earlier post seems like every one is complaining about the ratio. Here I have forced it to 1:1 . And I am using downloaded version dlib-19.4. Do I have to PULL it from git hub.

**Comment by davisking** [comment]
You don't have the current dlib if that's the message you get.  Here is the message in the current code: https://github.com/davisking/dlib/blob/master/dlib/svm/structural_svm_object_detection_problem.h#L267

It's not what you posted.

**Comment by kumarvis** [comment]
I just cloned it (https://github.com/davisking/dlib )and tried is that not enough ??  Not very experienced with github, Please help hope you dont mind .
